Combined license report for Java and Ruby dependencies

danhermann opened the pull request:
This work builds on the existing [Ruby script](https://github.com/elastic/logstash/pull/8837) that extracts license information for Ruby gems and generates a combined license report for both Ruby and Java dependencies. It consists of the following pieces executed in the following order:

- Invoke the existing Ruby dependencies script to generate that license report
- Invoke a new Gradle `generateLicenseReport` task for Java dependencies. The task performs the following operations:
  - Executes the [Gradle-License-Report](https://github.com/jk1/Gradle-License-Report) task on all subprojects. The Gradle-License-Report plugin does the heavy lifting of resolving all compile-time and runtime Java dependencies, transitive and otherwise, for any subprojects containing artifacts that are distributed (as defined by either the absence of the `isDistributedArtifact` property or `isDistributedArtifact=true` defined in the `gradle.properties` file) for each of them from Maven Central. It does also retrieve license data for each of them though we do not use that data as we have found it unreliable.
  - Executes a new Java tool called `dependencies-report` that takes as input the Ruby dependencies report and all the Java dependency reports. The tool then performs the following operations:
    - Combines, sorts, and de-dupes all the Ruby and Java dependencies
    - Locates the license for that dependency from a `licenseMapping.csv` file that is packaged with the tool.
    - Validates that the mapped license complies with our license policy by ensuring that the license exists in the `acceptableLicenses.csv` file that is packaged with the tool.
    - Writes the completed dependency license report to `$LOGSTASH_REPO_HOME/build/dependencyLicenseReport.csv`
    - If there were any dependencies for which a license did not exist in `licenseMapping.csv` or the mapped license did not comply with our license policy, the tool exits with a non-zero exit code.

andrewvc commented:
Running this locally I get:

```
C:\Users\andre\projects\logstash (HEAD detached at fa542b9 -> origin)
λ gradlew logstash-core:generateLicenseReport --stacktrace

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':logstash-core:generateLicenseReport'.
> Cannot access last() element from an empty List

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':logstash-core:generateLicenseReport'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:60)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:626)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:581)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:59)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:128)
        at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
        at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
        at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:46)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
        at org.gradle.initialization.DefaultGradleLauncher$ExecuteTasks.run(DefaultGradleLauncher.java:314)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.initialization.DefaultGradleLauncher.runTasks(DefaultGradleLauncher.java:204)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:134)
        at org.gradle.initialization.DefaultGradleLauncher.executeTasks(DefaultGradleLauncher.java:109)
        at org.gradle.internal.invocation.GradleBuildController$1.call(GradleBuildController.java:78)
        at org.gradle.internal.invocation.GradleBuildController$1.call(GradleBuildController.java:75)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:152)
        at org.gradle.internal.invocation.GradleBuildController.doBuild(GradleBuildController.java:100)
        at org.gradle.internal.invocation.GradleBuildController.run(GradleBuildController.java:75)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.tooling.internal.provider.ValidatingBuildActionRunner.run(ValidatingBuildActionRunner.java:32)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$1.run(RunAsBuildOperationBuildActionRunner.java:43)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner.run(RunAsBuildOperationBuildActionRunner.java:40)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionRunner.run(SubscribableBuildActionRunner.java:51)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:47)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:30)
        at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:39)
        at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:25)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:80)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:53)
        at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:57)
        at org.gradle.tooling.internal.provider.ServicesSetupBuildActionExecuter.execute(ServicesSetupBuildActionExecuter.java:32)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:36)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:25)
        at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:43)
        at org.gradle.tooling.internal.provider.ParallelismConfigurationBuildActionExecuter.execute(ParallelismConfigurationBuildActionExecuter.java:29)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:64)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:29)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:59)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:44)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:45)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:30)
        at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:51)
        at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:173)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:280)
        at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:253)
        at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:33)
        at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:24)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:33)
        at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:22)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:246)
        at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:182)
        at org.gradle.launcher.Main.doAction(Main.java:33)
        at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:60)
        at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:37)
        at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
        at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:30)
        at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:129)
        at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
Caused by: java.util.NoSuchElementException: Cannot access last() element from an empty List
        at com.github.jk1.license.render.LicenseDataCollector.singleModuleLicenseInfo(LicenseDataCollector.groovy:11)
        at com.github.jk1.license.render.CsvReportRenderer.renderDependency(CsvReportRenderer.groovy:64)
        at com.github.jk1.license.render.CsvReportRenderer$_render_closure1.doCall(CsvReportRenderer.groovy:59)
        at com.github.jk1.license.render.CsvReportRenderer.render(CsvReportRenderer.groovy:58)
        at com.github.jk1.license.render.ReportRenderer$render.call(Unknown Source)
        at com.github.jk1.license.ReportTask$_generateReport_closure3.doCall(ReportTask.groovy:42)
        at com.github.jk1.license.ReportTask.generateReport(ReportTask.groovy:41)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:784)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:751)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:121)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        ... 94 more


* Get more help at https://help.gradle.org

BUILD FAILED in 16s
1 actionable task: 1 executed

C:\Users\andre\projects\logstash (HEAD detached at fa542b9 -> origin)
λ
```

andrewvc commented on logstash-core/build.gradle:
```diff
@@ -15,12 +16,23 @@ repositories {
 
+apply plugin: 'com.github.jk1.dependency-license-report'
+
+licenseReport {
```
Shouldn't this depend on downloading a bunch of gems and setting up the env?

I think we probably need a task in the parent project that does those things then calls this.

andrewvc replied:
Doesn't this need an explicit `task` declaration? 

I'm not sure how to actually run this. Can you lend me some gradle expertise?

danhermann replied:
Yes, you can run this via `./gradlew logstash-core:generateLicenseReport`. The output will be written to the `logstash-core/build/reports/dependency-license` directory.

danhermann commented:
@andrewvc, I've updated the PR to include the version that works correctly with our gradle build.

andrewvc commented:
Hmmm, this mostly looks correct. It appears that the java licenses are in the form of URLs. @mgreau what is the requirement here? If we report `"https://www.apache.org/licenses/LICENSE-2.0.txt"` as a license is that acceptable, or do we need to alter it to become `Apache-2.0`?

danhermann commented:
Added root task to run license report for all subprojects -- `./gradlew generateLicenseReport`. Will consolidate individual subproject results into single report if @mgreau confirms that the existing format is acceptable.

mgreau commented:
Thanks for the ping.

We use [the license SPDX short identifier](https://spdx.org/licenses/) to be consistent across all projects, it means that we will recognize `Apache-2.0` as the SPDX identifier for Apache License 2.0 and then we will use this URL for the license text: 

- https://spdx.org/licenses/Apache-2.0.html

If you need to add a custom URL to point to a license not listed on SPDX, you can use `;` after the identifier to set the license URL, for example: 

`Custom;https://raw.githubusercontent.com/en-wl/wordlist/scowl-7.1/scowl/Copyright` or 
`Public-Domain;http://icon.shef.ac.uk/Moby/`

danhermann commented:
@andrewvc, if we need to provide SPDX identifiers for the licenses, I think a post-processing step will be necessary because neither Ruby Gems nor Maven packages require that licenses be specified using either the SPDX name or URL. I'd suggest we maintain a text file mapping of encountered license name/URLs to their SPDX counterparts. The post-processing step would match these up and produce a master list of dependency licenses in the desired format with SPDX identifiers. Any unknown licenses would be designed `UNKNOWN` and a warning would be printed.

Let me know if that is acceptable and if so, I'll start on it.

danhermann commented:
@andrewvc, could you let me know when you've determined whether this can be run at build time in a build environment or if it also needs to be able to be run after install time in a deployment environment?

andrewvc commented:
@danhermann based on our discussion yesterday, and the fact that beats does this AOT, I think we're fine there.

danhermann commented:
@andrewvc, this is now ready for review.

andrewvc commented:
jenkins, please test this.

andrewvc commented:
Running this command locally I'm not seeing a CSV generated:

```
~/p/logstash (danhermann-gradle_license_report) $ git rev-parse HEAD
2dd943caab198573475d59eb1b04640ee10b4af3
~/p/logstash (danhermann-gradle_license_report) $ bin/dependencies-report --csv report.csv
Finding gem dependencies
Finding java/jar dependencies
~/p/logstash (danhermann-gradle_license_report) $ cat gradle.properties ^C
~/p/logstash (danhermann-gradle_license_report) $ ls *report.csv*
```

andrewvc commented:
FYI, I've opened a PR to change the strategy to be fully manual.

danhermann commented:
@andrewvc, I merged your PR, removed all the code related to automated license detection, added code to validate that any manually-mapped license was a compliant license, and added two more unit tests to cover failure scenarios.

andrewvc commented on tools/dependencies-report/src/main/resources/licenseMapping.csv:
```diff
@@ -0,0 +1,45 @@
+"filewatch:0.9.0",MIT
+"gelfd:0.2.0",Apache-2.0
+"gmetric:0.1.3",MIT
+"UNKNOWNgzip:1.0",UNKNOWN
```
We can remove this. I've published a new version of the ES output that no longer depends on this.

andrewvc commented on bin/logstash.lib.sh:
```diff
@@ -167,3 +167,10 @@ ruby_exec() {
   exec "${JRUBY_BIN}" "$@"
 }
+
+ruby_run() {
```
I added this, but I wound up not using it. We should remove it from this PR.

andrewvc commented on tools/dependencies-report/src/main/resources/licenseMapping.csv:
```diff
@@ -0,0 +1,45 @@
+"gelfd:0.2.0",Apache-2.0
+"gmetric:0.1.3",MIT
+"UNKNOWNgzip:1.0",UNKNOWN
+"UNKNOWNjls-grok:0.11.4",UNKNOWN
```
We are blocked on this until @jordansissel adds a license to https://github.com/jordansissel/ruby-grok
